pkg/fuzzer/queue/queue.go

```diff
@@ -0,0 +1,375 @@
+		}
+	}
+	return nil
+}
+
+type callback struct {
+	cb func() *Request
+}
+
+// Callback produces a source that calls the callback to serve every Next() request.
+func Callback(cb func() *Request) Source {
+	return &callback{cb}
+}
+
+func (cb *callback) Next() *Request {
+	return cb.cb()
+}
+
+type alternate struct {
+	base Source
+	mu   sync.Mutex
+	rnd  *rand.Rand
+	prob float32
+}
+
+// Alternate returns a nil *Request in `prob` share of requests.
+func Alternate(base Source, rnd *rand.Rand, prob float32) Source {
+	return &alternate{
+		base: base,
+		rnd:  rnd,
+		prob: prob,
+	}
+}
+
+func (a *alternate) Next() *Request {
+	var skip bool
+	if a.mu.TryLock() {
+		skip = a.rnd.Float32() < a.prob
+		a.mu.Unlock()
+	}
+	if skip {
+		return nil
+	}
+	return a.base.Next()
+}
+
+type PriorityQueue struct {
+	mu       *sync.Mutex
+	ops      *priorityQueueOps[*Request]
+	currPrio priority
+}
+
+func Priority() *PriorityQueue {
+	return &PriorityQueue{
+		mu:       &sync.Mutex{},
+		ops:      &priorityQueueOps[*Request]{},
+		currPrio: priority{0},
+	}
+}
+
+func (pq *PriorityQueue) AppendQueue() *PriorityQueue {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	pq.currPrio = pq.currPrio.next()
+	nextPrio := append(priority{}, pq.currPrio...)
+	return &PriorityQueue{
+		// We use the same queue, therefore the same mutex.
+		mu:       pq.mu,
+		ops:      pq.ops,
+		currPrio: append(nextPrio, 0),
+	}
+}
+
+func (pq *PriorityQueue) Submit(req *Request) {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	pq.currPrio = pq.currPrio.next()
+	pq.ops.Push(req, pq.currPrio)
+}
+
+func (pq *PriorityQueue) Next() *Request {
+	pq.mu.Lock()
+	defer pq.mu.Unlock()
+	return pq.ops.Pop()
+}
+
+type DynamicMultiplexer struct {
+	mu   sync.Mutex
+	list []Source
+}
+
+func (dm *DynamicMultiplexer) Add(source Source) {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	dm.list = append(dm.list, source)
+}
+
+func (dm *DynamicMultiplexer) Next() *Request {
+	dm.mu.Lock()
+	list := dm.list
+	dm.mu.Unlock()
+	for _, source := range list {
+		req := source.Next()
+		if req != nil {
+			return req
+		}
+	}
+	return nil
+}
+
+// Deduplicator() keeps track of the previously run requests to avoid re-running them.
+type Deduplicator struct {
+	mu     sync.Mutex
+	ctx    context.Context
+	source Source
+	hash   func(*Request) hash.Sig
+	mm     map[hash.Sig]*Request
+}
+
+func NewDeduplicator(ctx context.Context, source Source, hashFunc func(*Request) hash.Sig) *Deduplicator {
+	return &Deduplicator{
+		ctx:    ctx,
+		source: source,
+		hash:   hashFunc,
+		mm:     map[hash.Sig]*Request{},
+	}
+}
+
+func (d *Deduplicator) Next() *Request {
+	for {
+		req := d.source.Next()
+		if req == nil {
+			return nil
+		}
+		hash := d.hash(req)
+		d.mu.Lock()
+		dupOf, ok := d.mm[hash]
+		if !ok {
+			d.mm[hash] = req
+		}
+		d.mu.Unlock()
+		if !ok {
+			// This is the first time we see such a request.
+			return req
+		}
+		// Transfer the result from the original request.
+		go func() {
+			res := dupOf.Wait(d.ctx)
```

Comment:
I don't like that because Deduplicator we complicate Request and make it more expensive (always create the chan + mutex).

Can we use Done callback wrapping here to achieve the same goal? It looks like it may provide a nice solution -- when we first return a request and add to d.mm, we wrap its Done callback to intercept completion.

Reply:
We already create a channel for every Request anyway, so this change shouldn't make things worse. 

> Can we use Done callback wrapping here to achieve the same goal?

I was thinking about that, and it should theoretically be doable, but then we'd have to make the `OnDone()` 
1) Thread-safe -- we don't know when Done() will be called on the `dupOf` request. So there will be a mutex again.
2) It must support adding callbacks after the Request is executed. E.g. just call the callback immediately. It somewhat breaks the semantics of the callback function -- now it returns true/false depending on whether we should keep on unwinding the stack of callbacks or we should stop.

Adding these temporary goroutines seemed much easier and less prone to errors. In the end, there will be just a few thousands of them and they won't stay too long.

Reply:
I don't think it needs to be thread-safe, nor we need to support adding OnDone after completion. It should be adder _before_ deduplicator returned the original request for execution (before it's returned from Next, it can't be done).

Reply:
> We already create a channel for every Request anyway, so this change shouldn't make things worse.

Ah, I see, we now don't know if we will wait for request or not when when we issue the request... it's a bit unfortunate. I liked the old Fuzzer.exec method which created the done channel for starting the request.

Reply:
I've rewritten the deduplicator code to rely on OnDone().